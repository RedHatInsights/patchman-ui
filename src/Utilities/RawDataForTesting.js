// raw data for testing purposes only

export const advisoryRows = [{
    attributes: {
        advisory_type: 1,
        applicable_systems: 1,
        description: 'The microcode_ctl ',
        public_date: '2020-06-24T17:22:25Z',
        synopsis: 'microcode_ctl',
        severity: 3
    },
    id: 'RHEA-2020:2743',
    type: 'advisory'
}];

export const systemAdvisoryRows = [{
    attributes: {
        advisory_type: 3,
        description: 'Kernel-based Virtual.',
        public_date: '2020-06-30T13:38:53Z',
        severity: 3,
        synopsis: 'Important'
    },
    id: 'RHSA-2020:2774',
    type: 'advisory'
}];

export const systemRows = [{
    attributes: {
        display_name: 'automation_host',
        last_evaluation: '2020-08-12T13:57:54.028883Z',
        last_upload: '2020-08-12T09:26:33.891907Z',
        packages_installed: 0,
        packages_updatable: 0,
        rhba_count: 213,
        rhea_count: 35,
        rhsa_count: 77,
        stale: false
    },
    id: '8ddb54f5-aeeb-49b1-8448-c29049d686c1',
    type: 'system'
}];

export const advisoryHeader = {
    attributes: {
        advisory_type: 3,
        description: 'Kernel-based Virtual.',
        public_date: '2020-06-30T13:38:53Z',
        modified_date: '2020-06-30T13:38:53Z',
        severity: 3,
        synopsis: 'Important'
    },
    id: 'RHSA-2020:2774',
    type: 'advisory'
};

export const advisoryDetailRows = {
    data: {
        attributes: {
            cves: ['CVE-2020-11080'],
            description: 'libnghttp2.',
            fixes: null,
            modified_date: '2020-06-25T16:31:29Z',
            packages: { libnghttp2: '1.33.0-3.el8_2.1.x86_64' },
            public_date: '2020-06-25T16:31:29Z',
            references: [],
            severity: 3,
            solution: 'For details',
            synopsis: 'Important: ',
            topic: 'An update .'
        }
    },
    id: 'RHSA-2020:2755',
    type: 'advisory'
};
export const systemPackages = [{
    description: 'This package contains testing data',
    evra: '2.2.*',
    name: 'acl',
    summary: 'Access control list utilities',
    updates: null
}];

export const entityDetail = {
    loaded: true,
    activeApps: [],
    entity: {
        account: '60'
    }
};
