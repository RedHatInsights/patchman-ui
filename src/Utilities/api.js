import { encodeApiParams } from './Helpers';

export function createApiCall(
    endpoint,
    method,
    parameters = undefined,
    data = undefined
) {
    if (parameters && method === 'get') {
        endpoint = endpoint.concat(encodeApiParams(parameters));
    }

    let result = window.insights.chrome.auth
    .getUser()
    .then(() =>
        fetch('/api/patch/v1' + endpoint, {
            method,
            credentials: 'include',
            body: JSON.stringify(data)
        })
    )
    .then(res => {
        if (!res.ok) {
            const contentType = res.headers.get('content-type');
            if (contentType.indexOf('json') !== -1) {
                throw res;
            } else {
                throw {
                    errors: [
                        { status: res.status, detail: res.statusText }
                    ]
                };
            }
        }

        return res.json();
    })
    .catch(caughtError => {
        const error = Promise.resolve(caughtError || {});
        const genericError = {
            title:
                    'There was an error getting data'
        };
        return error.then(error => {
            const [info] = error.errors;
            const result = error.errors && { ...genericError, detail: info.detail, status: info.status } || genericError;
            throw result ;
        });
    });
    return result;
}

export const fetchApplicableAdvisoriesApi = params => {
    return createApiCall('/advisories', 'get', params);
};

export const fetchApplicableSystemAdvisoriesApi = params => {
    let { id, ...allParams } = params;
    return createApiCall(`/systems/${id}/advisories`, 'get', allParams);
};

export const fetchSystems = params => {
    return createApiCall('/systems', 'get', params);
};

export const fetchAdvisoryDetailsApi = params => {
    return createApiCall(`/advisories/${params.advisoryName}`, 'get');
};

export const fetchApplicablePackagesApi = params => {
    let { id, ...allParams } = params;
    return createApiCall(`/systems/${id}/packages`, 'get', allParams);
};

export const fetchAffectedSystems = params => {
    const { id, ...args } = params;
    return createApiCall(`/advisories/${id}/systems`, 'get', args);
};

export const fetchPackagesList = params => {
    return createApiCall('/packages', 'get', params);
};

export const exportAdvisoriesCSV = params => {
    let endpoint = '/export/advisories';
    endpoint = endpoint.concat(encodeApiParams(params));
    return fetch('/api/patch/v1' + endpoint, {
        method: 'get',
        credentials: 'include',
        headers: new Headers({ accept: 'text/csv' })
    }).then(res => res.text());
};

export const exportAdvisoriesJSON = params => {
    let endpoint = '/export/advisories';
    endpoint = endpoint.concat(encodeApiParams(params));
    return fetch('/api/patch/v1' + endpoint, {
        method: 'get',
        credentials: 'include',
        headers: new Headers({ accept: 'application/json' })
    }).then(res => res.json());
};

export const exportSystemsCSV = params => {
    let endpoint = '/export/systems';
    endpoint = endpoint.concat(encodeApiParams(params));
    return fetch('/api/patch/v1' + endpoint, {
        method: 'get',
        credentials: 'include',
        headers: new Headers({ accept: 'text/csv' })
    }).then(res => res.text());
};

export const exportSystemsJSON = params => {
    let endpoint = '/export/systems';
    endpoint = endpoint.concat(encodeApiParams(params));
    return fetch('/api/patch/v1' + endpoint, {
        method: 'get',
        credentials: 'include',
        headers: new Headers({ accept: 'application/json' })
    }).then(res => res.json());
};
