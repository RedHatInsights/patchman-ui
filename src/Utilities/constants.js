import BugIcon from '@patternfly/react-icons/dist/js/icons/bug-icon';
import EnhancementIcon from '@patternfly/react-icons/dist/js/icons/enhancement-icon';
import SecurityIcon from '@patternfly/react-icons/dist/js/icons/security-icon';
import React from 'react';
import { subtractDate } from './Helpers';

export const STATUS_REJECTED = 'rejected';
export const STATUS_LOADING = 'loading';
export const STATUS_RESOLVED = 'resolved';

export const compoundSortValues = {
    operating_system: {
        asc: 'os_name,os_major,os_minor',
        desc: '-os_name,-os_major,-os_minor'
    },
    applicable_advisories: {
        asc: 'rhsa_count,rhba_count,rhea_count',
        desc: '-rhsa_count,-rhba_count,-rhea_count'
    }
};

// messy check because of frontend-components tests
export const ENABLE_PACKAGES = window?.insights?.chrome?.isBeta && insights?.chrome?.isBeta();
export const storeListDefaults = {
    rows: [],
    metadata: {
        limit: 25,
        offset: 0,
        total_items: 0
    },
    expandedRows: {},
    selectedRows: {},
    queryParams: {},
    error: {}
};

export const publicDateOptions = [
    { apiValue: '', label: 'All', value: 'all' },
    {
        apiValue: `gt:${subtractDate(7)}`,
        label: 'Last 7 days',
        value: 'last7'
    },
    {
        apiValue: `gt:${subtractDate(30)}`,
        label: 'Last 30 days',
        value: 'last30'
    },
    {
        apiValue: `gt:${subtractDate(90)}`,
        label: 'Last 90 days',
        value: 'last90'
    },
    {
        apiValue: `gt:${subtractDate(365)}`,
        label: 'Last year',
        value: 'lastYear'
    },
    {
        apiValue: `lt:${subtractDate(365)}`,
        label: 'More than 1 year ago',
        value: 'moreThanYear'
    }
];

export const advisorySeverities = [
    {
        value: 0,
        label: 'N/A',
        color: 'var(--pf-global--Color--200)'
    },
    {
        value: 1,
        label: 'Low',
        color: 'var(--pf-global--Color--200)',
        text:
            'This rating is given to all other issues that have a security impact. These are the types of vulnerabilities' +
            ' that are believed to require unlikely circumstances to be able to be exploited, or where a successful exploit' +
            ' would give minimal consequences.'
    },
    {
        value: 2,
        label: 'Moderate',
        color: 'var(--pf-global--warning-color--100)',
        text:
            'This rating is given to flaws that may be more difficult to exploit but could still lead to some' +
            ' compromise of the confidentiality, integrity, or availability of resources, under certain circumstances.' +
            ' These are the types of vulnerabilities that could have had a Critical impact or Important impact' +
            ' but are less easily exploited based on a technical evaluation of the flaw, or affect unlikely' +
            ' configurations.'
    },
    {
        value: 3,
        label: 'Important',
        color: '#ec7a08',
        text:
            'This rating is given to flaws that can easily compromise the confidentiality, integrity, or availability' +
            ' of resources. These are the types of vulnerabilities that allow local users to gain privileges, allow' +
            ' unauthenticated remote users to view resources that should otherwise be protected by authentication,' +
            ' allow authenticated remote users to execute arbitrary code, or allow remote users to cause a denial' +
            ' of service.'
    },
    {
        value: 4,
        label: 'Critical',
        color: 'var(--pf-global--danger-color--100)',
        text:
            'This rating is given to flaws that could be easily exploited by a remote unauthenticated attacker' +
            ' and lead to system compromise (arbitrary code execution) without requiring user interaction.' +
            ' These are the types of vulnerabilities that can be exploited by worms. Flaws that require an authenticated' +
            ' remote user, a local user, or an unlikely configuration are not classed as Critical impact.'
    }
];

export const advisoryTypes = [
    {
        value: 3,
        label: 'Security',
        icon: <SecurityIcon />
    },
    {
        value: 2,
        label: 'Bugfix',
        icon: <BugIcon />
    },
    {
        value: 1,
        label: 'Enhancement',
        icon: <EnhancementIcon />
    }
];

export const updatableTypes = [
    {
        value: '',
        label: 'All'
    },
    {
        value: true,
        label: 'Upgradable'
    }
];

export const packagesListUpdatableTypes = [
    {
        value: '',
        label: 'All'
    },
    {
        value: 'gt:0',
        label: 'Upgradable'
    }
];

export const filterCategories = {
    advisory_type: {
        label: 'Advisory type',
        values: advisoryTypes
    },
    public_date: {
        label: 'Public date',
        values: publicDateOptions
    },
    updatable: {
        label: 'Status',
        values: updatableTypes
    },
    systems_updatable: {
        label: 'Status',
        values: packagesListUpdatableTypes
    }
};

export const entityTypes = {
    advisories: 'advisories',
    packages: 'packages'
};

export const ReadOnlyNotification = {
    title: 'title',
    detail: 'message'
};

export const remediationIdentifiers = {
    package: 'patch-package',
    advisory: 'patch-advisory'
};
