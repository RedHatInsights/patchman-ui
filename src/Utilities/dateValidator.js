import { intl } from './IntlProvider';
import messages from '../Messages';

/*
validates date of a type YYYY-MM-DD with starting limit 1990-01-01.
regex is used to comply the Patternfly date format.
*/
const dateValidator = (dateStr) => {
    const regex = /^(\d{4})-(0[1-9]|1[012])-(0?[1-9]|[12][0-9]|3[01])$/;

    if (dateStr && typeof dateStr === 'string' && dateStr.match(regex) === null) {
        return intl.formatMessage(messages.labelsErrorInvalidDate);
    }

    const date = new Date(dateStr);
    const timestamp = date.getTime();

    //Month had to be set to equal 0 to get the first month of the year.
    const minDate = new Date(1990, 0, 1);

    if (typeof timestamp !== 'number' || Number.isNaN(timestamp)) {
        return intl.formatMessage(messages.labelsErrorInvalidDate);
    }

    if (date < minDate) {
        return intl.formatMessage(messages.labelsErrorDateLimit);
    }

    return;
};

export default dateValidator;
