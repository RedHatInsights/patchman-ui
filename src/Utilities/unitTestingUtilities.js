import { useSelector } from 'react-redux';

/* eslint-disable */
export const initMocks = () => {

    window.insights = {
        loadInventory: (args) => new Promise((resolve, reject) =>  {
            resolve(({
                inventoryConnector: (args) => {
                            const InventoryTable = ({ children }) => <div >A mock passed! {children} </div>;
                            return ({ InventoryTable });
                        },
                        mergeWithEntities: () => {}
            }));
        }),
        loadRemediations: () => new Promise((resolve, reject) =>  {
            resolve({ 
                openWizard: () => new Promise((resolve, reject) =>  { 
                    resolve(true)
                })
            });
        })
    };

};

export const mockStore = (initialState, mutatedState) => {
    const customMiddleWare = store => next => action => {
        useSelector.mockImplementation(callback => {
            return callback(mutatedState);
        });
        next(action);
    };

    const mockStore = configureStore([customMiddleWare]);

    return mockStore(initialState);
}
/* eslint-enable */
