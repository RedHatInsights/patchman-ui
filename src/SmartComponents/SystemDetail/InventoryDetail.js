import React, { useEffect } from 'react';
import { Main } from '@redhat-cloud-services/frontend-components/Main';
import { useSelector, useDispatch } from 'react-redux';
import Header from '../../PresentationalComponents/Header/Header';
import { paths } from '../../Routes';
import { register } from '../../store';
import { SystemDetailStore } from '../../store/Reducers/SystemDetailStore';
import { intl } from '../../Utilities/IntlProvider';
import messages from '../../Messages';
import { setPageTitle } from '../../Utilities/Hooks';
import { InventoryDetailHead, AppInfo, DetailWrapper } from '@redhat-cloud-services/frontend-components/Inventory';
import { Alert, Grid, GridItem, TextContent, Text } from '@patternfly/react-core';
import { fetchSystemDetailsAction } from '../../store/Actions/Actions';
import propTypes from 'prop-types';
import { clearNotifications } from '@redhat-cloud-services/frontend-components-notifications/redux';;
import PatchSetWizard from '../PatchSetWizard/PatchSetWizard';

const InventoryDetail = ({ match }) => {

    const [patchSetState, setBaselineState] = React.useState({
        isOpen: false,
        shouldRefresh: false,
        systemsIDs: []
    });

    const dispatch = useDispatch();
    const entityDetails = useSelector(
        ({ entityDetails }) => entityDetails && entityDetails.entity
    );

    const { hasThirdPartyRepo, patchSetName } = useSelector(
        ({ entityDetails }) => entityDetails ?? {}
    );
    const entityId = match.params?.inventoryId;
    useEffect(() => {
        dispatch(fetchSystemDetailsAction(entityId));
        return () => {
            dispatch(clearNotifications());
        };
    }, [patchSetState.shouldRefresh]);

    const pageTitle = entityDetails && `${entityDetails.display_name} - ${intl.formatMessage(messages.titlesSystems)}`;
    setPageTitle(pageTitle);

    const showBaselineModal = () => {
        setBaselineState({ isOpen: true, systemsIDs: [entityId] });
    };

    return (
        <DetailWrapper
            onLoad={({ mergeWithDetail }) => {
                register({
                    ...mergeWithDetail(SystemDetailStore)
                });
            }}
        >
            {(patchSetState.isOpen) &&
                <PatchSetWizard systemsIDs={patchSetState.systemsIDs} setBaselineState={setBaselineState} />}
            <Header
                title=""
                headerOUIA={'inventory-details'}
                breadcrumbs={[
                    {
                        title: intl.formatMessage(messages.titlesPatchSystems),
                        to: paths.systems.to,
                        isActive: false
                    },
                    entityDetails && {
                        title: entityDetails.display_name,
                        isActive: true
                    }
                ]}
            >
                <InventoryDetailHead hideBack
                    showTags
                    actions={[
                        {
                            title: intl.formatMessage(messages.titlesPatchSetAssignMultipleButton),
                            key: 'assign-to-patch-set',
                            onClick: showBaselineModal
                        }]}
                >
                    <Grid>
                        <GridItem>
                            {patchSetName && <TextContent>
                                <Text>
                                    {`${intl.formatMessage(messages.labelsColumnsPatchSet)}: ${patchSetName}`}
                                </Text>
                            </TextContent>
                            }
                        </GridItem>
                        <GridItem>
                            {hasThirdPartyRepo &&
                                (<Alert className='pf-u-mt-md' isInline variant="info"
                                    title={intl.formatMessage(messages.textThirdPartyInfo)}>
                                </Alert>)
                            }
                        </GridItem>
                    </Grid>
                </InventoryDetailHead>
            </Header>
            <Main>
                <AppInfo />
            </Main>
        </DetailWrapper>
    );
};

InventoryDetail.propTypes = {
    match: propTypes.object
};

export default InventoryDetail;
