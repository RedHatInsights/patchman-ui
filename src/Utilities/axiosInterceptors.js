import {
    authInterceptor,
    responseDataInterceptor
} from '@redhat-cloud-services/frontend-components-utilities/files/interceptors';
import axios from 'axios';
import { ReadOnlyNotification } from './constants';
const axiosInstance = axios.create();

export async function  accessCheckInterceptor (config)  {

    const userPermissions = await insights.chrome.getUserPermissions();
    const inventoryPermissionList = ['inventory:*:*', 'inventory:*:read', 'inventory:hosts:read'];

    const hasInventoryAccess = userPermissions.some((access) => inventoryPermissionList.includes(access?.permission));

    if (!hasInventoryAccess) {
        throw new axios.Cancel('Operation canceled. User does not have access to Inventory');
    }

    return config;
}

export function errorInterceptor(err) {

    if (!axios.isCancel(err)) {

        const { response, isAxiosError } = { ...err };

        if (response && isAxiosError) {
            const { status, statusText, data } = response;

            if (!status) {
                throw err;
            } else {
                const genericError = {
                    title:
                            'There was an error getting data'
                };

                const result = { ...genericError, detail: data.error && data.error || statusText, status };

                throw result;
            }
        }

        throw err;
    }
}

export function readOnlyInterceptor(error) {
    if (error.response && error.response.status === 503) {
        const data = ReadOnlyNotification;
        throw data;
    }

    throw error;
}

axiosInstance.interceptors.request.use(accessCheckInterceptor);
axiosInstance.interceptors.request.use(authInterceptor);
axiosInstance.interceptors.response.use(responseDataInterceptor);
axiosInstance.interceptors.response.use(null, readOnlyInterceptor);
axiosInstance.interceptors.response.use(null, errorInterceptor);

export default axiosInstance;
