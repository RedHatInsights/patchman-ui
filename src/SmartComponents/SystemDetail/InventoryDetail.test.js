import InventoryDetail from './InventoryDetail';
import toJson from 'enzyme-to-json';
import { Provider } from 'react-redux';
import { entityDetail } from '../../Utilities/RawDataForTesting';
import configureStore from 'redux-mock-store';
import { initMocks } from '../../Utilities/unitTestingUtilities.js';
import { BrowserRouter as Router } from 'react-router-dom';
import { useSelector } from 'react-redux';

initMocks();

jest.mock('../../Utilities/Hooks', () => ({
    ...jest.requireActual('../../Utilities/Hooks'),
    useFeatureFlag: jest.fn()
}));

jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux'),
    useSelector: jest.fn()
}));

jest.mock('@redhat-cloud-services/frontend-components/Inventory', () => ({
    ...jest.requireActual('@redhat-cloud-services/frontend-components/Inventory'),
    InventoryTable: jest.fn(() => <div className='testInventroyComponent'><div>This is child</div></div>),
    InventoryDetailHead: jest.fn(() => <div className='testInventoryDetailHead'><div>This is child</div></div>),
    AppInfo: jest.fn(() => <div className='testInventroyAppInfo'><div>This is child</div></div>),
    DetailWrapper: jest.fn(({ children }) => <div className='testDetailWrapper'><div>{children}</div></div>)
}));

const mockState = { ...entityDetail };

const initStore = (state) => {
    const customMiddleWare = () => next => action => {
        useSelector.mockImplementation(callback => {
            return callback({  entityDetails: state });
        });
        next(action);
    };

    const mockStore = configureStore([customMiddleWare]);
    return mockStore({  entityDetails: state });
};

let wrapper;
let store = initStore(mockState);

beforeEach(() => {
    console.error = () => {};

    store.clearActions();
    useSelector.mockImplementation(callback => {
        return callback({ entityDetails: mockState });
    });
    wrapper = mount(<Provider store={store}>
        <Router><InventoryDetail match = {{ params: { inventoryId: 'test' } }}/></Router>
    </Provider>);
});

afterEach(() => {
    useSelector.mockClear();
});

describe('InventoryPage.js', () => {
    it('Should match the snapshots', () => {
        expect(toJson(wrapper.update())).toMatchSnapshot();
    });
});

