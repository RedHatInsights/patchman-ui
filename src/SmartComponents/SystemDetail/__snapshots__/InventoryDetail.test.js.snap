// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InventoryPage.js Should match the snapshots 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <InventoryDetail
        match={
          Object {
            "params": Object {
              "inventoryId": "test",
            },
          }
        }
      >
        <mockConstructor
          onLoad={[Function]}
        >
          <div
            className="testDetailWrapper"
          >
            <div>
              <Header
                breadcrumbs={
                  Array [
                    Object {
                      "isActive": false,
                      "title": "Patch systems",
                      "to": "/systems/",
                    },
                    Object {
                      "isActive": true,
                      "title": undefined,
                    },
                  ]
                }
                headerOUIA="inventory-details"
                title=""
              >
                <PageHeader
                  data-ouia-component-type="inventory-details-page-header"
                >
                  <section
                    className="pf-l-page-header pf-c-page-header pf-l-page__main-section pf-c-page__main-section pf-m-light"
                    data-ouia-component-type="inventory-details-page-header"
                    widget-type="InsightsPageHeader"
                  >
                    <HeaderBreadcrumbs
                      headerOUIA="inventory-details"
                      items={
                        Array [
                          Object {
                            "isActive": false,
                            "title": "Patch systems",
                            "to": "/systems/",
                          },
                          Object {
                            "isActive": true,
                            "title": undefined,
                          },
                        ]
                      }
                    >
                      <Breadcrumb>
                        <nav
                          aria-label="Breadcrumb"
                          className="pf-c-breadcrumb"
                          data-ouia-component-id="OUIA-Generated-Breadcrumb-1"
                          data-ouia-component-type="PF4/Breadcrumb"
                          data-ouia-safe={true}
                        >
                          <ol
                            className="pf-c-breadcrumb__list"
                          >
                            <BreadcrumbItem
                              isActive={false}
                              key=".$Patch systems"
                              showDivider={false}
                            >
                              <li
                                className="pf-c-breadcrumb__item"
                              >
                                <Link
                                  data-ouia-component-id="breadcrumb-to-Patch systems"
                                  data-ouia-component-type="inventory-details-breadcrumb"
                                  to="/systems/"
                                >
                                  <LinkAnchor
                                    data-ouia-component-id="breadcrumb-to-Patch systems"
                                    data-ouia-component-type="inventory-details-breadcrumb"
                                    href="/systems/"
                                    navigate={[Function]}
                                  >
                                    <a
                                      data-ouia-component-id="breadcrumb-to-Patch systems"
                                      data-ouia-component-type="inventory-details-breadcrumb"
                                      href="/systems/"
                                      onClick={[Function]}
                                    >
                                      Patch systems
                                    </a>
                                  </LinkAnchor>
                                </Link>
                              </li>
                            </BreadcrumbItem>
                            <BreadcrumbItem
                              isActive={true}
                              key=".1"
                              showDivider={true}
                            >
                              <li
                                className="pf-c-breadcrumb__item"
                              >
                                <span
                                  className="pf-c-breadcrumb__item-divider"
                                >
                                  <AngleRightIcon
                                    color="currentColor"
                                    noVerticalAlign={false}
                                    size="sm"
                                  >
                                    <svg
                                      aria-hidden={true}
                                      aria-labelledby={null}
                                      fill="currentColor"
                                      height="1em"
                                      role="img"
                                      style={
                                        Object {
                                          "verticalAlign": "-0.125em",
                                        }
                                      }
                                      viewBox="0 0 256 512"
                                      width="1em"
                                    >
                                      <path
                                        d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                                      />
                                    </svg>
                                  </AngleRightIcon>
                                </span>
                              </li>
                            </BreadcrumbItem>
                          </ol>
                        </nav>
                      </Breadcrumb>
                    </HeaderBreadcrumbs>
                    <PageHeaderTitle
                      title=""
                    >
                      <Title
                        className=""
                        headingLevel="h1"
                        size="2xl"
                        widget-type="InsightsPageHeaderTitle"
                      >
                        <h1
                          className="pf-c-title pf-m-2xl"
                          data-ouia-component-id="OUIA-Generated-Title-1"
                          data-ouia-component-type="PF4/Title"
                          data-ouia-safe={true}
                          widget-type="InsightsPageHeaderTitle"
                        />
                      </Title>
                    </PageHeaderTitle>
                    <mockConstructor
                      actions={
                        Array [
                          Object {
                            "key": "assign-to-patch-set",
                            "onClick": [Function],
                            "title": "Assign to patch sets",
                          },
                        ]
                      }
                      hideBack={true}
                      showTags={true}
                    >
                      <div
                        className="testInventoryDetailHead"
                      >
                        <div>
                          This is child
                        </div>
                      </div>
                    </mockConstructor>
                  </section>
                </PageHeader>
              </Header>
              <Connect(InternalMain)>
                <InternalMain>
                  <section
                    className="pf-l-page__main-section pf-c-page__main-section"
                    page-type=""
                  >
                    <mockConstructor>
                      <div
                        className="testInventroyAppInfo"
                      >
                        <div>
                          This is child
                        </div>
                      </div>
                    </mockConstructor>
                  </section>
                </InternalMain>
              </Connect(InternalMain)>
            </div>
          </div>
        </mockConstructor>
      </InventoryDetail>
    </Router>
  </BrowserRouter>
</Provider>
`;
